HW2 is completed (base)

SelferRy opened the pull request:
Solved basic version of hw02 task.

Commit:
HW2 is completed (base)

Commit:
refactor(HW2):  encapsulate Atoi to isDigit

Commit:
feat(HW2):  imp for backslash cases

SelferRy commented:
Added backslash part

engineer-go commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,120 @@
+	}
+	fmt.Println(res)
+	fmt.Println(expected)
+}
```
main вам тут ни к чему - для этогого есть тесты
```suggestion

```

engineer-go commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,120 @@
+func isDigit(r rune) bool {
+	_, err := strconv.Atoi(string(r))
+	return err == nil
+}
```
тут вы вызываете strconv.Atoi который преобразует (конвертирует) строку в число. оверхед невелик, но имхо понятнее было бы  <= rune <=

SelferRy replied:
Переделал

engineer-go commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,120 @@
+	return res.String(), nil
+}
+
+func firstIsDigit(first rune) bool {
```
нарушен DRY - полностью повторяет isDigit

SelferRy replied:
Переделал

engineer-go commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,120 @@
+	return err == nil
+}
+
+func lastElem(i int, count int) bool {
```
предлагаю заинлайнить

SelferRy replied:
1. Функция выделяет логический блок - помогает не вникать каждый раз в низкоуровневый код, который не говорит о сути происходящего. 
2. Функция объясняет что происходит своим названием, позволяет не думать о несущественных деталях на том уровне абстракции, где используется.
3. Функцию можно явно протестировать.
4. Функция расположена локально в данном модуле и является приватной - т.е. не увеличивает как-либо значимо сложность кода и плюсы достаются "бесплатно".

Считаю, что лучше оставить, т.к. повышает читабельность и согласованность данного модуля.

engineer-go commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,120 @@
+	return err == nil
+}
+
+func isBackslash(r rune) bool {
```
то же самое что и в функциях выше, oneliner который понятнее было бы использовать внутри unpack

SelferRy replied:
Могу повторить аргументацию из блока выше и добавить, что:
5. Данная функция неоднократно переиспользуется, что подчёркивает общность + протестировав её однажды гарантируем правильное выполнение каждого блока (ошибки встречаются и в банальных вещах).

engineer-go commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,120 @@
-package hw02unpackstring
+package main
```
название модуля менять нельзя

SelferRy replied:
Вернул

engineer-go commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,120 @@
+	return string(r) == "\\"
+}
+
+func takeTimes(num rune) int {
```
то же замечание - можно убрать функцию, она не несёт полезных функций

SelferRy replied:
Могу повторить аргументы выше и добавить:
6. Данная функция менее тривиальна, чем предыдущие.
7. В местах, где она используется, мне не нужны детали: какая внешняя либа, ошибка, как всё записывается и т.п.
8. Данная функция содержит зависимость от библиотеки (пусть и стандартной), которая скрывается.

engineer-go commented on hw02_unpack_string/unpack_test.go:
```diff
@@ -1,4 +1,4 @@
-package hw02unpackstring
+package main
```
название модуля менять нельзя

SelferRy replied:
Вернул

engineer-go commented on hw02_unpack_string/unpack_test.go:
```diff
@@ -36,7 +35,6 @@ func TestUnpack(t *testing.T) {
 func TestUnpackInvalidString(t *testing.T) {
 	invalidStrings := []string{"3abc", "45", "aaa10b"}
```
посмотрите, почему некорректно отрабатывает следующий тест кейс "ghhj\\б2"

SelferRy replied:
Посмотрел. Добавил в тесты

engineer-go commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,120 @@
+			times := takeTimes(s[i+1])
+			res.WriteString(strings.Repeat(string(s[i]), times))
+			i++
+		case matchBackslash(s[i], s[i+1]):
```
вложенные case - switch очень сложны для понимания.  В идеале было бы обойтись без него. в целом тут достаточно нескольких if - else

SelferRy replied:
Вынес данную логику. 
Преимущество в читабельности if перед switch в случае > 2 альтернатив нет. 
Могут быть плюсы в механике (когда надо неск условий проверить и сделать что-то не по 1, а по неск условиям и действия независимы - это тоже, кстати, усложняет чтение). Или чтобы переменные ввести для/на время проверки.
В данном случае эти механики не используются и switch используется как более легковесная и прямолинейная альтернатива if + if else + else.

engineer-go reviewed (changes requested):
Набросал замечания, посмотрите пожалуйста как будет время

SelferRy committed:
Update hw02_unpack_string/unpack.go

Co-authored-by: ityulkanov <ityulkanov@gmail.com>

Commit:
refactor(HW2):  encapsulate backslashStrategy, define backslash-states, change isDigit and matchBackslash

Commit:
test(HW2):  upg functional tests, introduce tests for sub-func.

Commit:
refactor(HW2):  collapse backslash-digit cases, extract strategy, reuse matchBackslash

Commit:
refactor(HW2):
- extract isThirdDigit cond,
- rename isDigitAfter,
- clarify isMultiplyBackslash,isMultiplyDigit

Commit:
refactor(HW2): dry many backslashes case

Commit:
refactor(HW2): make help-func pure, IO in Unpack

Commit:
refactor(HW2): extract isForbidden and dry

Commit:
test(HW2): add edge cases

Commit:
test(HW2): add multiply "\n" case

Commit:
fix(HW2): linter

SelferRy commented:
> Набросал замечания, посмотрите пожалуйста как будет время

ответил по замечаниям
